core-bundle/tests/Security/Authentication/AuthenticationFailureHandlerTest.php

```diff
@@ -130,12 +131,17 @@ private function getRequest(): Request
             ->willReturn($session)
         ;
 
-        $request->request = $this->createMock(ParameterBag::class);
-        $request->request
-            ->expects($this->never())
-            ->method('get')
-            ->with('username')
-        ;
+        $params = [
+            'username' => new class() implements \Stringable {
+                public function __toString(): string
+                {
+                    AuthenticationFailureHandlerTest::fail('Unexpected call to get("username")');
```

Comment:
This will not lead to an assertion that the test was successful if the method is never called.

Reply:
If the method was never called the test is successful. What do you want to change?

Reply:
The test does not fail, but it does not assert a successful test. The test suite does not know whether you actually tested something successfully.

PS: you shouldn't resolve a conversation if you ask me a question 😆

Reply:
> The test does not fail, but it does not assert a successful test

There are still 3 assertions in this test. Besides that triggering `self::fail()` is equivalent to`self::never()` in a mocked method AFAIK. 

There is an edge case where you do neither have any positive assertions, nor any uses of `never()` - there PHPUnit cannot interpret the call to `fail()` and would (falsely) complain that the test doesn't make assertions. But that's a limitation of PHPUnit and IMHO not the case here.

Reply:
> PS: you shouldn't resolve a conversation if you ask me a question 😆

I probably sensed that you will not answer it anyways 😜

> The test does not fail, but it does not assert a successful test.

Unable to reproduce 
![Bildschirmfoto 2021-07-28 um 13 01 09](https://user-images.githubusercontent.com/367169/127311801-0f7e464d-9ed3-41b7-bd0e-7b47e9805a6c.png)

Reply:
yeah, those are three other assertions. There is no assertion that username is never retrieved. I think there must be another way to write a better test 🤷

Reply:
> I think there must be another way to write a better test

Feel free to create a PR to replace all `fail()` calls with something else, but I don’t see how this is a bad test.

Reply:
A simple solution in general would be to check a variable outside the callback: https://stackoverflow.com/questions/9296529/phpunit-how-to-test-if-callback-gets-called

However, I checked the code, and there seems to be absolutely not reason for this check at all (most likely became obsolete). The request is never queried for any property, and the username is not relevant at all. The `logException` method does not even use `$request`, so I'd suggest to simply remove the method argument and the test 🙃

Reply:
Still unrelated to this PR IMO. Anyway, removed in 544ddfd5a712b802fba996d4c952f861b4ead024